File: sudoq-app/sudoqmodel/src/test/java/de/sudoq/model/game/GameTests.java
package de.sudoq.model.game;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import org.junit.BeforeClass;
import org.junit.Test;

import de.sudoq.model.Utility;
import de.sudoq.model.actionTree.ActionTreeElement;
import de.sudoq.model.actionTree.NoteActionFactory;
import de.sudoq.model.actionTree.SolveActionFactory;
import de.sudoq.model.profile.Profile;
import de.sudoq.model.solverGenerator.Generator;
import de.sudoq.model.solverGenerator.GeneratorCallback;
import de.sudoq.model.solverGenerator.solution.Solution;
import de.sudoq.model.sudoku.Cell;
import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.PositionMap;
import de.sudoq.model.sudoku.Sudoku;
import de.sudoq.model.sudoku.SudokuBuilder;
import de.sudoq.model.sudoku.complexity.Complexity;
import de.sudoq.model.sudoku.sudokuTypes.SudokuTypes;
import de.sudoq.model.sudoku.sudokuTypes.TypeBuilder;

public class GameTests {

	private static Sudoku sudoku;

	@BeforeClass
	public static void beforeClass() {
		Utility.copySudokus();
		File profileDir = new File("/tmp/sudoq/GameTests/profile");
		Profile.Companion.getInstance(profileDir);

		TypeBuilder.get99(); //just to force initialization of filemanager
		
		GeneratorCallback gc = new GeneratorCallback() {
			@Override
			public void generationFinished(Sudoku sudoku) {
				GameTests.sudoku = sudoku;
			}

			@Override
			public void generationFinished(Sudoku sudoku, List<Solution> sl) {
				GameTests.sudoku = sudoku;
			}
		};
		
		new Generator().generate(SudokuTypes.standard9x9, Complexity.easy, gc);
	}

	@Test
	public void testInstanciation() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());
		assertEquals(game.getId(), 2);
		assertNotNull(game.getStateHandler());
		assertEquals(game.getSudoku().getCell(Position.get(8, 8)).getCurrentValue(), Cell.EMPTYVAL);
		assertNull(game.getSudoku().getCell(Position.get(10, 2)));
		assertEquals(0, game.getAssistancesCost());
	}

	@Test(expected = NullPointerException.class)
	public void testNullInstantiation() {
		new Game(2, null);
	}

	@Test
	public void testGameInteraction() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());

		Position pos = Position.get(1, 1);
		ActionTreeElement start = game.getCurrentState();
        Cell f = game.getSudoku().getCell(pos);
		game.addAndExecute(new SolveActionFactory().createAction(3, f));//setze 3
		game.addAndExecute(new SolveActionFactory().createAction(4, f));//setze 4
		assertFalse(game.isMarked(game.getCurrentState()));
		game.addAndExecute(new SolveActionFactory().createAction(5, f));//setze 5
		assertEquals(5, game.getSudoku().getCell(pos).getCurrentValue());
		game.markCurrentState();
		assertTrue(game.isMarked(game.getCurrentState()));

		game.goToState(start);
		assertEquals(Cell.EMPTYVAL, f.getCurrentValue());
		assertFalse(game.isFinished());

		game.redo();
		game.redo();
		assertEquals(5, f.getCurrentValue());//schlägt fehl
		game.undo();
		assertEquals(Cell.EMPTYVAL, f.getCurrentValue());
		game.redo();
		assertFalse(game.checkSudoku());

		game.addTime(23);
		assertEquals(game.getTime(), 23);
	}

	@Test
	public void testEquals() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());
		assertEquals(game, game);
		Game game2 = new Game(3, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());
		assertNotEquals(game, game2);

		Position pos = Position.get(1, 1);
		ActionTreeElement start = game.getCurrentState();

		game.addAndExecute(new SolveActionFactory().createAction(3, game.getSudoku().getCell(pos)));
		game.addAndExecute(new SolveActionFactory().createAction(4, game.getSudoku().getCell(pos)));
		game.addAndExecute(new SolveActionFactory().createAction(5, game.getSudoku().getCell(pos)));
		game.markCurrentState();

		game.goToState(start);

		game.redo();
		game.redo();
		game.undo();

		assertNotEquals(game, game2);
		assertEquals(game, game);

		assertNotEquals(game, new Object());
	}

	@Test
	public void testGameXML() {
		Sudoku s = new SudokuBuilder(SudokuTypes.standard9x9).createSudoku();
		s.setId(5);
		Game game = new Game(2, s);

		Position pos = Position.get(1, 1);
		ActionTreeElement start = game.getCurrentState();

		game.addAndExecute(new SolveActionFactory().createAction(3, game.getSudoku().getCell(pos)));
		game.addAndExecute(new SolveActionFactory().createAction(4, game.getSudoku().getCell(pos)));
		game.addAndExecute(new SolveActionFactory().createAction(5, game.getSudoku().getCell(pos)));
		game.addAndExecute(new NoteActionFactory().createAction(2, game.getSudoku().getCell(pos)));
		assertEquals(game.getSudoku().getCell(pos).getCurrentValue(), 5);
		game.markCurrentState();
		assertTrue(game.isMarked(game.getCurrentState()));

		game.goToState(start);

		game.redo();
		game.redo();
		game.undo();

		Game game2 = new Game();
		game2.fillFromXml(game.toXmlTree());
		assertEquals(game, game2);
	}

	// Regression Test for Issue-89
	@Test
	public void testFinishedAttributeConsistency() {
		SudokuBuilder sb = new SudokuBuilder(SudokuTypes.standard9x9);
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				sb.addSolution(Position.get(i, j), 1);
			}
		}
		Game game = new Game(1, sb.createSudoku());
		assertTrue(game.solveAll());
		assertTrue(game.isFinished());

		Game game2 = new Game();
		game2.fillFromXml(game.toXmlTree());
		assertTrue(game2.isFinished());

		game2.undo();
		assertTrue(game2.isFinished());

		game = new Game();
		game.fillFromXml(game2.toXmlTree());
		assertTrue(game.isFinished());
	}

	@Test(expected = NullPointerException.class)
	public void testSetNullAssistances() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());
		game.setAssistances(null);
	}

	@Test
	public void testAssistanceSetting() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());

		game.setAssistances(new GameSettings() {
			@Override
			public boolean getAssistance(Assistances assistance) {
				return true;
			}
		});

		for (Assistances a : Assistances.values()) {
			assertTrue(game.isAssistanceAvailable(a));
		}
	}

	@Test
	public void testHelp() {
		class SudokuMock extends Sudoku {
			private boolean finished = false;
			private boolean errors = false;

			public SudokuMock() {
				super(SudokuBuilder.createType(SudokuTypes.standard9x9), new PositionMap<Integer>(Position.get(9, 9)),
						new PositionMap<Boolean>(Position.get(9, 9)));
			}

			@Override
			public boolean isFinished() {
				return finished;
			}

			@Override
			public boolean hasErrors() {
				return errors;
			}

			public void toogleErrors() {
				errors = !errors;
			}

			public void toogleFinished() {
				finished = !finished;
			}

		}
		SudokuMock sudoku = new SudokuMock();
		Game game = new Game(2, sudoku);

		assertFalse(game.solveCell(null));
		assertFalse(game.solveCell(new Cell(true, -1, 3, 9)));

		game.addAndExecute(new SolveActionFactory().createAction(2, game.getSudoku().getCell(Position.get(0, 0))));
		sudoku.toogleErrors();
		assertFalse(game.checkSudoku());
		assertTrue(game.getCurrentState().isMistake());
		assertFalse(game.solveCell());
		assertFalse(game.solveCell(game.getSudoku().getCell(Position.get(0, 0))));
		assertFalse(game.solveAll());

		sudoku.toogleErrors();
		sudoku.toogleFinished();
		game.addAndExecute(new SolveActionFactory().createAction(Cell.EMPTYVAL,
				game.getSudoku().getCell(Position.get(0, 0))));
		game.addAndExecute(new SolveActionFactory().createAction(1, game.getSudoku().getCell(Position.get(0, 0))));
		assertEquals(Cell.EMPTYVAL, game.getSudoku().getCell(Position.get(0, 0)).getCurrentValue());
	}

	@Test
	public void testNoteAdjustment() {
		Game game = new Game(2, new SudokuBuilder(SudokuTypes.standard9x9).createSudoku());
		GameSettings as = new GameSettings();
		as.setAssistance(Assistances.autoAdjustNotes);
		game.setAssistances(as);

		game.addAndExecute(new NoteActionFactory().createAction(2, game.getSudoku().getCell(Position.get(1, 0))));
		game.addAndExecute(new NoteActionFactory().createAction(3, game.getSudoku().getCell(Position.get(1, 0))));
		game.addAndExecute(new NoteActionFactory().createAction(2, game.getSudoku().getCell(Position.get(0, 1))));
		game.addAndExecute(new NoteActionFactory().createAction(3, game.getSudoku().getCell(Position.get(0, 1))));

		assertTrue(game.getSudoku().getCell(Position.get(0, 1)).isNoteSet(2));
		assertTrue(game.getSudoku().getCell(Position.get(1, 0)).isNoteSet(2));
		assertTrue(game.getSudoku().getCell(Position.get(0, 1)).isNoteSet(3));
		assertTrue(game.getSudoku().getCell(Position.get(1, 0)).isNoteSet(3));

		game.addAndExecute(new SolveActionFactory().createAction(2, game.getSudoku().getCell(Position.get(0, 0))));

		assertTrue(game.getSudoku().getCell(Position.get(0, 1)).isNoteSet(3));
		assertTrue(game.getSudoku().getCell(Position.get(1, 0)).isNoteSet(3));
		assertFalse(game.getSudoku().getCell(Position.get(0, 1)).isNoteSet(2));
		assertFalse(game.getSudoku().getCell(Position.get(1, 0)).isNoteSet(2));
	}

	@Test
	public void testScore() {
		for (Complexity c : Complexity.values()) {
			Sudoku sudoku = new Sudoku(TypeBuilder.get99());
			sudoku.setComplexity(c);
			Game game = new Game(0, sudoku);
			game.addTime(60);
			switch (c) {
			case easy:
				assertEquals(game.getScore(), 2430);
				break;
			case medium:
				assertEquals(game.getScore(), 7290);
				break;
			case difficult:
				assertEquals(game.getScore(), 21870);
				break;
			case infernal:
				assertEquals(game.getScore(), 65610);
				break;
			}
		}
	}

	@Test
	public synchronized void testSolve() {
		System.out.println("before while");
		
		int counter = 0;
		while (GameTests.sudoku == null && counter<120) {
			try {
				wait(1000);
				counter++;
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		if(GameTests.sudoku == null)
			throw new IllegalStateException("infinite loop!");
		
        System.out.println("we passed the while loop!");
		Game game = new Game(1, sudoku);
		ArrayList<Cell> unsolvedCells = new ArrayList<Cell>();
		for (Cell f : sudoku) {
			if (f.isEditable()) {
				f.clearCurrentValue();
				unsolvedCells.add(f);
			}
		}
		assertTrue(game.checkSudoku());
		assertTrue(game.solveCell());
		boolean hasNewSolved = false;
		for (Cell f : sudoku) {
			if (f.isEditable() && f.isSolvedCorrect()) {
				if (hasNewSolved) {
					fail("Solve field solved more than one field");
				} else {
					hasNewSolved = true;
					unsolvedCells.remove(f);
				}
			}
		}
		assertTrue(hasNewSolved);
		assertTrue(game.checkSudoku());
		assertFalse(game.isFinished());
		assertTrue(game.solveCell(unsolvedCells.get(0)));
		assertTrue(unsolvedCells.get(0).isSolvedCorrect());
		unsolvedCells.remove(0);
		assertFalse(game.isFinished());
		assertTrue(game.solveAll());
		for (Cell f : unsolvedCells) {
			assertTrue(f.isSolvedCorrect());
		}
		assertTrue(game.isFinished());
		assertTrue(game.isFinished());
	}

	@Test
	public synchronized void testGoToLastCorrectState() {
		while (sudoku == null) {
			try {
				wait(100);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		Game game = new Game(1, sudoku);

		Cell unsolvedCell = null;
		for (Cell f : sudoku) {
			if (f.isEditable()) {
				f.clearCurrentValue();
				if (unsolvedCell == null)
					unsolvedCell = f;
			}
		}

		int oldAssistanceCost = game.getAssistancesCost();
		game.goToLastCorrectState();
		assertTrue(game.getAssistancesCost() - 3 == oldAssistanceCost);
		oldAssistanceCost = game.getAssistancesCost();

		if (unsolvedCell.getSolution() < 8) {
			game.addAndExecute(new SolveActionFactory().createAction(8, unsolvedCell));
		} else {
			game.addAndExecute(new SolveActionFactory().createAction(7, unsolvedCell));
		}

		game.goToLastCorrectState();
		assertTrue(game.getAssistancesCost() - 3 == oldAssistanceCost);
		oldAssistanceCost = game.getAssistancesCost();
		assertTrue(unsolvedCell.getCurrentValue() == Cell.EMPTYVAL);
		assertTrue(game.getCurrentState().isCorrect());
	}

	// Regression Test for Issue-90
	@Test
	public void testAutoAdjustNotesForAutomaticSolving() {
		SudokuBuilder sb = new SudokuBuilder(SudokuTypes.standard9x9);
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				sb.addSolution(Position.get(i, j), 1);
			}
		}
		Game game = new Game(2, sb.createSudoku());
		GameSettings as = new GameSettings();
		as.setAssistance(Assistances.autoAdjustNotes);
		game.setAssistances(as);

		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				game.addAndExecute(new NoteActionFactory().createAction(1, game.getSudoku().getCell(Position.get(i, j))));
				assertTrue(game.getSudoku().getCell(Position.get(i, j)).isNoteSet(1));
			}
		}

		assertTrue(game.solveCell());
		boolean done = false;
		int x = -1;
		int y = -1;
		for (int i = 0; i < 9 && !done; i++) {
			for (int j = 0; j < 9 && !done; j++) {
				if (game.getSudoku().getCell(Position.get(i, j)).getCurrentValue() == 1) {
					done = true;
					x = i;
					y = j;
				}
			}
		}
		assertTrue(done);

		for (int i = 0; i < 9; i++) {
			assertFalse(game.getSudoku().getCell(Position.get(x, i)).isNoteSet(1));
			assertFalse(game.getSudoku().getCell(Position.get(i, y)).isNoteSet(1));
		}
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				assertFalse(game.getSudoku().getCell(Position.get((x - (x % 3) + i), (y - (y % 3) + j))).isNoteSet(1));
			}
		}

		x = (x + 3) % 9;
		y = (y + 3) % 9;
		game.solveCell(game.getSudoku().getCell(Position.get(x, y)));

		for (int i = 0; i < 9; i++) {
			assertFalse(game.getSudoku().getCell(Position.get(x, i)).isNoteSet(1));
			assertFalse(game.getSudoku().getCell(Position.get(i, y)).isNoteSet(1));
		}
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				assertFalse(game.getSudoku().getCell(Position.get((x - (x % 3) + i), (y - (y % 3) + j))).isNoteSet(1));
			}
		}
	}
}

File: sudoq-app/sudoqmodel/src/test/kotlin/de/sudoq/model/game/GameTests.kt
package de.sudoq.model.game

import de.sudoq.model.actionTree.NoteActionFactory
import de.sudoq.model.actionTree.SolveActionFactory
import de.sudoq.model.solverGenerator.Generator
import de.sudoq.model.solverGenerator.GeneratorCallback
import de.sudoq.model.solverGenerator.solution.Solution
import de.sudoq.model.solverGenerator.utils.PrettySudokuRepo2
import de.sudoq.model.solverGenerator.utils.SudokuTypeRepo4Tests
import de.sudoq.model.sudoku.Cell
import de.sudoq.model.sudoku.Position
import de.sudoq.model.sudoku.PositionMap
import de.sudoq.model.sudoku.Sudoku
import de.sudoq.model.sudoku.SudokuBuilder
import de.sudoq.model.sudoku.complexity.Complexity
import de.sudoq.model.sudoku.sudokuTypes.SudokuTypeProvider.getSudokuType
import de.sudoq.model.sudoku.sudokuTypes.SudokuTypes
import de.sudoq.model.sudoku.sudokuTypes.TypeBuilder
import org.amshove.kluent.invoking
import org.amshove.kluent.`should be`
import org.amshove.kluent.`should be equal to`
import org.amshove.kluent.`should not be`
import org.amshove.kluent.`should not be equal to`
import org.amshove.kluent.`should throw`
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.nio.file.Paths
import java.time.Duration
import java.time.temporal.ChronoUnit
import java.util.concurrent.CompletableFuture

internal class GameTests {
    private val sudokuRepo = PrettySudokuRepo2(sudokuTypeRepo)

    /**
     * This is the former @BeforeAll init method. it takes suspiciously long to generate a sudoku todo investigate!
     */
    @Test
    @Throws(Exception::class)
    fun debugGeneration() {
        TypeBuilder.get99() //just to force initialization of filemanager

        val future = CompletableFuture<Sudoku?>()

        val gc: GeneratorCallback = object : GeneratorCallback {
            override fun generationFinished(sudoku: Sudoku) {
                future.complete(sudoku)
            }

            override fun generationFinished(sudoku: Sudoku, sl: List<Solution>) {
                future.complete(sudoku)
            }
        }
        Generator(sudokuTypeRepo).generate(SudokuTypes.standard9x9, Complexity.easy, gc)
        Assertions.assertTimeoutPreemptively(Duration.of(60, ChronoUnit.SECONDS)) {
            future.get()
        }
        val sudoku = future.get()
        sudoku `should not be` null
    }

    @Test
    fun instantiation() {
        val game = Game(2, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())
        game.id `should be equal to` 2
        game.stateHandler `should not be` null
        game.sudoku!!.getCell(Position[8, 8])!!.currentValue `should be equal to` Cell.EMPTYVAL
        game.sudoku!!.getCell(Position[10, 2]) `should be` null
        game.assistancesCost `should be equal to` 0
    }

    @Test
    fun gameInteraction() {
        val game = Game(2, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())

        val pos = Position[1, 1]
        val start = game.currentState
        val f = game.sudoku!!.getCell(pos)
        game.addAndExecute(SolveActionFactory().createAction(3, f!!)) //setze 3
        game.addAndExecute(SolveActionFactory().createAction(4, f)) //setze 4
        game.isMarked(game.currentState) `should be` false
        game.addAndExecute(SolveActionFactory().createAction(5, f)) //setze 5
        game.sudoku!!.getCell(pos)!!.currentValue `should be equal to` 5
        game.markCurrentState()
        game.isMarked(game.currentState) `should be` true

        game.goToState(start)
        f.currentValue `should be equal to` Cell.EMPTYVAL
        game.isFinished() `should be` false

        game.redo()
        game.redo()
        f.currentValue `should be equal to` 5 //schlägt fehl
        game.undo()
        f.currentValue `should be equal to` Cell.EMPTYVAL
        game.redo()
        game.checkSudoku() `should be` false

        game.addTime(23)
        game.time `should be equal to` 23
    }

    @Test
    fun equals() {
        val game = Game(2, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())
        game `should be equal to` game
        val game2 = Game(3, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())
        game2 `should not be equal to` game

        val pos = Position[1, 1]
        val start = game.currentState

        game.addAndExecute(SolveActionFactory().createAction(3, game.sudoku!!.getCell(pos)!!))
        game.addAndExecute(SolveActionFactory().createAction(4, game.sudoku!!.getCell(pos)!!))
        game.addAndExecute(SolveActionFactory().createAction(5, game.sudoku!!.getCell(pos)!!))
        game.markCurrentState()

        game.goToState(start)

        game.redo()
        game.redo()
        game.undo()

        game2 `should not be equal to` game
        game `should be equal to` game

        game `should not be equal to` Any()
    }

    @Test
    fun gameXML() {
        val s = SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku()
        s.id = 5
        val game = Game(2, s)

        val pos = Position[1, 1]
        val start = game.currentState

        game.addAndExecute(SolveActionFactory().createAction(3, game.sudoku!!.getCell(pos)!!))
        game.addAndExecute(SolveActionFactory().createAction(4, game.sudoku!!.getCell(pos)!!))
        game.addAndExecute(SolveActionFactory().createAction(5, game.sudoku!!.getCell(pos)!!))
        game.addAndExecute(NoteActionFactory().createAction(2, game.sudoku!!.getCell(pos)!!))
        game.sudoku!!.getCell(pos)!!.currentValue `should be equal to` 5
        game.markCurrentState()
        game.isMarked(game.currentState) `should be` true

        game.goToState(start)

        game.redo()
        game.redo()
        game.undo()
    }

    // Regression Test for Issue-89
    @Test
    fun finishedAttributeConsistency() {
        val sb = SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo)
        for (i in 0..8) {
            for (j in 0..8) {
                sb.addSolution(Position[i, j], 1)
            }
        }
        val game = Game(1, sb.createSudoku())
        game.solveAll() `should be` true
        game.isFinished() `should be` true
    }

    @Test
    fun assistanceSetting() {
        val game = Game(2, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())

        game.setAssistances(object : GameSettings() {
            override fun getAssistance(assistance: Assistances): Boolean = true
        })

        Assistances.entries
            .map(game::isAssistanceAvailable)
            .forEach(Assertions::assertTrue)
    }

    @Test
    fun help() {
        class SudokuMock : Sudoku(
            getSudokuType(SudokuTypes.standard9x9, sudokuTypeRepo), PositionMap(Position[9, 9]),
            PositionMap(Position[9, 9])
        ) {
            override var isFinished: Boolean = false
                private set
            private var errors = false

            public override fun hasErrors(): Boolean {
                return errors
            }

            fun toogleErrors() {
                errors = !errors
            }

            fun toogleFinished() {
                this.isFinished = !this.isFinished
            }
        }

        val sudoku = SudokuMock()
        val game = Game(2, sudoku)

        game.solveCell(null) `should be` false
        game.solveCell(Cell(true, -1, 3, 9)) `should be` false

        game.addAndExecute(SolveActionFactory().createAction(2, game.sudoku!!.getCell(Position[0, 0])!!))
        sudoku.toogleErrors()
        game.checkSudoku() `should be` false
        game.currentState.isMistake `should be` true
        game.solveCell() `should be` false
        game.solveCell(game.sudoku!!.getCell(Position[0, 0])) `should be` false
        game.solveAll() `should be` false

        sudoku.toogleErrors()
        sudoku.toogleFinished()
        game.addAndExecute(
            SolveActionFactory().createAction(
                Cell.EMPTYVAL,
                game.sudoku!!.getCell(Position[0, 0])!!
            )
        )
        game.addAndExecute(SolveActionFactory().createAction(1, game.sudoku!!.getCell(Position[0, 0])!!))
        game.sudoku!!.getCell(Position[0, 0])!!.currentValue `should be equal to` Cell.EMPTYVAL
    }

    @Test
    fun noteAdjustment() {
        val game = Game(2, SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo).createSudoku())
        val `as` = GameSettings()
        `as`.setAssistance(Assistances.autoAdjustNotes)
        game.setAssistances(`as`)

        game.addAndExecute(NoteActionFactory().createAction(2, game.sudoku!!.getCell(Position[1, 0])!!))
        game.addAndExecute(NoteActionFactory().createAction(3, game.sudoku!!.getCell(Position[1, 0])!!))
        game.addAndExecute(NoteActionFactory().createAction(2, game.sudoku!!.getCell(Position[0, 1])!!))
        game.addAndExecute(NoteActionFactory().createAction(3, game.sudoku!!.getCell(Position[0, 1])!!))

        game.sudoku!!.getCell(Position[0, 1])!!.isNoteSet(2) `should be` true
        game.sudoku!!.getCell(Position[1, 0])!!.isNoteSet(2) `should be` true
        game.sudoku!!.getCell(Position[0, 1])!!.isNoteSet(3) `should be` true
        game.sudoku!!.getCell(Position[1, 0])!!.isNoteSet(3) `should be` true

        game.addAndExecute(SolveActionFactory().createAction(2, game.sudoku!!.getCell(Position[0, 0])!!))

        game.sudoku!!.getCell(Position[0, 1])!!.isNoteSet(3) `should be` true
        game.sudoku!!.getCell(Position[1, 0])!!.isNoteSet(3) `should be` true
        game.sudoku!!.getCell(Position[0, 1])!!.isNoteSet(2) `should be` false
        game.sudoku!!.getCell(Position[1, 0])!!.isNoteSet(2) `should be` false
    }

    @ParameterizedTest
    @CsvSource(value = ["easy,2430", "medium,7290", "difficult,21870", "infernal,65610"])
    fun score(c: Complexity, score: Int) {
        val sudoku = Sudoku(TypeBuilder.get99())
        sudoku.complexity = c
        val game = Game(0, sudoku)
        game.addTime(60)
        game.score `should be equal to` score
    }

    @Test
    fun `score for arbitrary should not be possible`() {
        val sudoku = Sudoku(TypeBuilder.get99())
        sudoku.complexity = Complexity.arbitrary
        val game = Game(0, sudoku)
        game.addTime(60)
        invoking { game.score } `should throw` IllegalStateException::class
    }

    @Test
    @Synchronized
    fun solve() {
        val sudokuPath = Paths.get("sudokus/9_lockedCandidates_1.pretty")
        val sudoku = sudokuRepo.read(sudokuPath, Complexity.easy)
        sudoku `should not be` null
        val game = Game(1, sudoku)
        val unsolvedCells = ArrayList<Cell>()
        for (f in sudoku) {
            if (f.isEditable) {
                f.clearCurrentValue()
                unsolvedCells.add(f)
            }
        }
        game.checkSudoku() `should be` true
        game.solveCell() `should be` true
        var hasNewSolved = false
        for (f in sudoku) {
            if (f.isEditable && f.isSolvedCorrect) {
                if (hasNewSolved) {
                    Assertions.fail<Any>("Solve field solved more than one field")
                } else {
                    hasNewSolved = true
                    unsolvedCells.remove(f)
                }
            }
        }
        hasNewSolved `should be` true
        game.checkSudoku() `should be` true
        game.isFinished() `should be` false
        game.solveCell(unsolvedCells[0]) `should be` true
        unsolvedCells[0].isSolvedCorrect `should be` true
        unsolvedCells.removeAt(0)
        game.isFinished() `should be` false
        game.solveAll() `should be` true
        for (f in unsolvedCells) {
            f.isSolvedCorrect `should be` true
        }
        game.isFinished() `should be` true
        game.isFinished() `should be` true
    }

    @Test
    @Synchronized
    fun goToLastCorrectState() {
        val sudokuPath = Paths.get("sudokus/9_lockedCandidates_1.pretty")
        val sudoku = sudokuRepo.read(sudokuPath, Complexity.easy)
        sudoku `should not be` null
        val game = Game(1, sudoku)

        var unsolvedCell: Cell? = null
        for (f in sudoku) {
            if (f.isEditable) {
                f.clearCurrentValue()
                if (unsolvedCell == null) unsolvedCell = f
            }
        }

        var oldAssistanceCost = game.assistancesCost
        game.goToLastCorrectState()
        game.assistancesCost `should be equal to` oldAssistanceCost + 3
        oldAssistanceCost = game.assistancesCost

        if (unsolvedCell!!.solution < 8) {
            game.addAndExecute(SolveActionFactory().createAction(8, unsolvedCell))
        } else {
            game.addAndExecute(SolveActionFactory().createAction(7, unsolvedCell))
        }

        game.goToLastCorrectState()
        game.assistancesCost `should be equal to` oldAssistanceCost + 3
        oldAssistanceCost = game.assistancesCost
        unsolvedCell.currentValue `should be equal to` Cell.EMPTYVAL
        game.currentState.isCorrect `should be` true
    }

    // Regression Test for Issue-90
    @Test
    fun autoAdjustNotesForAutomaticSolving() {
        val sb = SudokuBuilder(SudokuTypes.standard9x9, sudokuTypeRepo)
        for (i in 0..8) {
            for (j in 0..8) {
                sb.addSolution(Position[i, j], 1)
            }
        }
        val game = Game(2, sb.createSudoku())
        val `as` = GameSettings()
        `as`.setAssistance(Assistances.autoAdjustNotes)
        game.setAssistances(`as`)

        for (pos in game.sudoku!!.sudokuType.validPositions) {
            game.addAndExecute(NoteActionFactory().createAction(1, game.sudoku!!.getCell(pos)!!))
            game.sudoku!!.getCell(pos)!!.isNoteSet(1) `should be` true
        }

        game.solveCell() `should be` true
        var done = false
        var x = -1
        var y = -1
        run {
            var i = 0
            while (i < 9 && !done) {
                var j = 0
                while (j < 9 && !done) {
                    if (game.sudoku!!.getCell(Position[i, j])!!.currentValue == 1) {
                        done = true
                        x = i
                        y = j
                    }
                    j++
                }
                i++
            }
        }
        done `should be` true

        for (i in 0..8) {
            game.sudoku!!.getCell(Position[x, i])!!.isNoteSet(1) `should be` false
            game.sudoku!!.getCell(Position[i, y])!!.isNoteSet(1) `should be` false
        }
        for (i in 0..2) {
            for (j in 0..2) {
                game.sudoku!!.getCell(
                    Position[
                        (x - (x % 3) + i),
                        (y - (y % 3) + j)
                    ]
                )!!.isNoteSet(1) `should be` false
            }
        }

        x = (x + 3) % 9
        y = (y + 3) % 9
        game.solveCell(game.sudoku!!.getCell(Position[x, y]))

        for (i in 0..8) {
            game.sudoku!!.getCell(Position[x, i])!!.isNoteSet(1) `should be` false
            game.sudoku!!.getCell(Position[i, y])!!.isNoteSet(1) `should be` false
        }
        for (i in 0..2) {
            for (j in 0..2) {
                game.sudoku!!.getCell(
                    Position[
                        (x - (x % 3) + i),
                        (y - (y % 3) + j)
                    ]
                )!!.isNoteSet(1) `should be` false
            }
        }
    }

    companion object {
        private val sudokuTypeRepo = SudokuTypeRepo4Tests()
    }
}
